crawler.rb

```diff
@@ -1,39 +1,64 @@
+require 'pry'
+require 'set'
+require 'JSON'
 require 'nokogiri'
 require 'open-uri'
-require 'JSON'
-require 'pry'
 require_relative './document'
-require_relative './document_collection'
 require_relative './url_fetcher'
 
 class Crawler
-	def initialize(document_collection, url_collection)
-		@document_collection = document_collection
-		@url_collection = url_collection
-	end
-
-	def done_crawling?
-		@url_collection.empty?
-	end
-
-	def crawl_next_url
-		url = @url_collection.next_url
-		content = UrlFetcher.fetch(url)
-		document = Document.new(url, content)
-		@document_collection.add_document(document)
-		puts document.domain_hrefs.take(5)
-		document.domain_hrefs.each do |href|
-			@url_collection.add_url(href)
-		end
-	end
-
-	def run(times_to_run)
-		times_to_run.times do
-			if done_crawling? 
-				puts "Crawling complete!"
-				exit 0
-			end
-			crawl_next_url
-		end
-	end
+  attr_accessor :url, :href, :connection
+  def initialize(connection)
+    @connection = connection 
+  end
+
+  def done_crawling?
+    @connection.execute("SELECT count(*) FROM URLS_to_crawl WHERE state = 'uncrawled'")[0][0] == 0
+  end
+
+  def crawl_next_url
+    begin
+      @connection.transaction
```

Comment:
Why not just do

```ruby
@connection.transaction do
  '... things to be put in a transaction here ...'
end
```

If your code fails to throw a `StandardError` or `SQLite3::Exception`, your transaction will not rollback. Doing it within the block ensures that the `rollback` will occur. You may be asking yourself "When will this ever happen?". TBH I dunno, but I'd rather not have to think about it and would prefer to use their safer block-supporting syntax to take care of it for me :smile:

If you REAAAALLLLY wanna know, then crack open the source for rails and see how they implement it. There's no better way to understand it than to peel back the magic and see what they're doing. 99% of the time you'll be like "Oh, that's it?", but that's ok. Sometimes you'll be like "Oh god why???". That's never fun.

`StandardError` is basically a catch-all too, so there's really no reason to have both `StandardError` and `SQLite3::Exception` catching and running the same block.